// appium-ios-driver/build/lib/cookies.js
/*
 * derived from jQuery Cookie Plugin v1.4.1
 * https://github.com/carhartl/jquery-cookie
 */

// needed to communicate/translate between JSONWire cookies and regular JavaScript cookies

'use strict';

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _appiumLogger = require('appium-logger');

var log = (0, _appiumLogger.getLogger)('Cookie');

// parses the value if needed and converts the value if a converter is provided
// internal function, not exported
function convertCookie(value, converter) {
  if (value.indexOf('"') === 0) {
    // this is a quoted cookied according to RFC2068
    // remove enclosing quotes and internal quotes and backslashes
    value = value.slice(1, -1).replace(/\\"/g, '"').replace(/\\\\/g, '\\');
  }

  var parsedValue = undefined;
  try {
    parsedValue = decodeURIComponent(value.replace(/\+/g, ' '));
  } catch (e) {
    // no need to fail if we can't decode
    log.warn(e);
  }

  return converter ? converter(parsedValue) : parsedValue;
}

// takes arguments given and creates a JavaScript Cookie
function createJSCookie(key, value) {
  var options = arguments.length <= 2 || arguments[2] === undefined ? {} : arguments[2];

  return [encodeURIComponent(key), '=', value, options.expires ? '; expires=' + options.expires : '', options.path ? '; path=' + options.path : '', options.domain ? '; domain=' + options.domain : '', options.secure ? '; secure' : ''].join('');
}

// takes the JavaScript cookieString and translates it into a JSONWire formatted cookie
function createJWPCookie(key, cookieString) {
  var converter = arguments.length <= 2 || arguments[2] === undefined ? null : arguments[2];

  var result = {};
  var cookies = cookieString ? cookieString.split('; ') : [];
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = _getIterator(cookies), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var cookie = _step.value;

      var parts = cookie.split('=');

      // get the first and second element as name and value
      var _name = decodeURIComponent(parts.shift());
      var val = parts[0];

      // if name is key, this is the central element of the cookie, so add as `name`
      // otherwise it is an optional element
      if (key && key === _name) {
        result.name = key;
        result.value = convertCookie(val, converter);
      } else {
        result[_name] = convertCookie(val, converter);
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator['return']) {
        _iterator['return']();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  return result;
}

// takes a JavaScript cookiestring and parses it for the value given the key
function getValue(key, cookieString) {
  var converter = arguments.length <= 2 || arguments[2] === undefined ? null : arguments[2];

  var result = createJWPCookie(key, cookieString, converter);

  // if `key` is undefined we want the entire cookie
  return _lodash2['default'].isUndefined(key) ? result : result.value;
}

// returns a cookie that expires on 01 Jan 1970
// assign the returned cookie to an existing cookie to delete that cookie
function expireCookie(key, options) {
  // override `expires` in `options`, and then make the cookie
  return createJSCookie(key, '', _lodash2['default'].assign({}, options, {
    expires: 'Thu, 01 Jan 1970 00:00:00 GMT'
  }));
}

// export individually and also (as default) as an object
exports.createJSCookie = createJSCookie;
exports.createJWPCookie = createJWPCookie;
exports.getValue = getValue;
exports.expireCookie = expireCookie;
exports['default'] = { createJSCookie: createJSCookie, createJWPCookie: createJWPCookie, getValue: getValue, expireCookie: expireCookie };
