// appium-xcuitest-driver/build/lib/real-device-management.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _teen_process = require('teen_process');

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

function getConnectedDevices() {
  var _ref, stdout, msg;

  return _regeneratorRuntime.async(function getConnectedDevices$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.prev = 0;
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap((0, _teen_process.exec)('idevice_id', ['-l']));

      case 3:
        _ref = context$1$0.sent;
        stdout = _ref.stdout;
        return context$1$0.abrupt('return', stdout.trim().split('\n'));

      case 8:
        context$1$0.prev = 8;
        context$1$0.t0 = context$1$0['catch'](0);

        if (!(context$1$0.t0.message.indexOf('Command \'idevice_id\' not found') === -1)) {
          context$1$0.next = 12;
          break;
        }

        throw context$1$0.t0;

      case 12:
        msg = 'The \'idevice_id\' program is not installed. If you are running ' + 'a real device test it is necessary. Install with \'brew install ' + 'libimobiledevice --HEAD\'';

        _logger2['default'].warn(msg);
        return context$1$0.abrupt('return', []);

      case 15:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[0, 8]]);
}

function resetRealDevice(device, opts) {
  var bundleId;
  return _regeneratorRuntime.async(function resetRealDevice$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(opts.bundleId && opts.fullReset)) {
          context$1$0.next = 18;
          break;
        }

        bundleId = opts.bundleId;

        _logger2['default'].debug('Full reset requested. Will try to uninstall the app \'' + bundleId + '\'.');
        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(device.isInstalled(bundleId));

      case 5:
        if (context$1$0.sent) {
          context$1$0.next = 8;
          break;
        }

        _logger2['default'].debug('App not installed. No need to uninstall');
        return context$1$0.abrupt('return');

      case 8:
        context$1$0.prev = 8;
        context$1$0.next = 11;
        return _regeneratorRuntime.awrap(device.remove(bundleId));

      case 11:
        context$1$0.next = 17;
        break;

      case 13:
        context$1$0.prev = 13;
        context$1$0.t0 = context$1$0['catch'](8);

        _logger2['default'].error('Could not remove \'' + bundleId + '\' from device');
        throw context$1$0.t0;

      case 17:
        _logger2['default'].debug('Removed \'' + bundleId + '\'');

      case 18:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[8, 13]]);
}

function runRealDeviceReset(device, opts) {
  return _regeneratorRuntime.async(function runRealDeviceReset$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(!opts.noReset || opts.fullReset)) {
          context$1$0.next = 7;
          break;
        }

        _logger2['default'].debug('Running ios real device reset flow');

        if (opts.noReset) {
          context$1$0.next = 5;
          break;
        }

        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(resetRealDevice(device, opts));

      case 5:
        context$1$0.next = 8;
        break;

      case 7:
        _logger2['default'].debug("Reset not set, continuing");

      case 8:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

exports.getConnectedDevices = getConnectedDevices;
exports.runRealDeviceReset = runRealDeviceReset;
