// appium-remote-debugger/build/test/helpers/webkit-remote-debugger-server.js
require('source-map-support').install();

'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _http = require('http');

var _http2 = _interopRequireDefault(_http);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _ws = require('ws');

var _ws2 = _interopRequireDefault(_ws);

var _appiumLogger = require('appium-logger');

var log = (0, _appiumLogger.getLogger)('RemoteDebugger');
var WebSocketServer = _ws2['default'].Server;

// fake server for allowing both http requests and
// websocket requests, as needed to test Appium's
// version of webkit remote debugging support

var WebKitRemoteDebuggerServer = (function () {
  function WebKitRemoteDebuggerServer() {
    _classCallCheck(this, WebKitRemoteDebuggerServer);

    this.server = null;
    this.websocketServer = null;
    this.nextResponse = null;
  }

  // start the server
  // if a websocket server is needed, pass in `true`

  _createClass(WebKitRemoteDebuggerServer, [{
    key: 'start',
    value: function start() {
      var ws = arguments.length <= 0 || arguments[0] === undefined ? false : arguments[0];
      return _regeneratorRuntime.async(function start$(context$2$0) {
        var _this = this;

        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (ws) {
              context$2$0.next = 4;
              break;
            }

            return context$2$0.abrupt('return', new _bluebird2['default'](function (resolve) {
              _this.server = _http2['default'].createServer(function (req, res) {
                res.writeHead(200, { 'Content-Type': 'application/json' });
                if (_this.nextResponse) {
                  res.end(JSON.stringify(_this.nextResponse));
                  _this.nextResponse = null;
                } else {
                  res.end(JSON.stringify({ id: 2, type: 'real' }));
                }
              });
              _this.server.listen(1337, 'localhost', resolve);
              log.debug('Server running at http://localhost:1337/');
            }));

          case 4:
            // need a fake websocket server
            // but it doesn't need to do anything but connect and disconnect
            this.ws = true;
            return context$2$0.abrupt('return', new _bluebird2['default'](function (resolve) {
              _this.server = new WebSocketServer({ host: 'localhost', port: 1337 }, resolve);
            }));

          case 6:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }

    // stop one or both of the servers.
  }, {
    key: 'stop',
    value: function stop() {
      return _regeneratorRuntime.async(function stop$(context$2$0) {
        var _this2 = this;

        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (this.ws) {
              context$2$0.next = 4;
              break;
            }

            return context$2$0.abrupt('return', new _bluebird2['default'](function (resolve) {
              if (_this2.server) {
                _this2.server.close(function (err) {
                  resolve('Stopped listening: ' + err);
                });
              } else {
                resolve('Not listening.');
              }
            }));

          case 4:
            // websocket server isn't asynchronous
            this.server.close();
            return context$2$0.abrupt('return', _bluebird2['default'].resolve());

          case 6:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }

    // set what the next call to the http server will respond with
  }, {
    key: 'respondWith',
    value: function respondWith(response) {
      this.nextResponse = response;
    }
  }]);

  return WebKitRemoteDebuggerServer;
})();

exports.WebKitRemoteDebuggerServer = WebKitRemoteDebuggerServer;

// just need a simple http server for non-websocket calls
